### ISSUE TITLE
Problem with plugin

### ISSUE THREAD

**Issue by ScarletEmanu** [archlinux/archinstall#3021]
There are some bugs in file "archinstall/lib/plugins.py" and "archinstall/__init__.py". 

# First problem #

If I add `--plugin https://raw.githubusercontent.com/phisch/archinstall-aur/master/archinstall-aur.py` in archinstall command, this error appears:

```
Loading plugin from url ParseResult(scheme='https', netloc='raw.githubusercontent.com', path='/phisch/archinstall-aur/refs/heads/master/archinstall-aur.py', params='', query='', fragment='')
Traceback (most recent call last):
  File "/home/archi/MyArchIso/install.py", line 3, in <module>
    from my_global_menu import Menu
  File "/home/archi/MyArchIso/my_global_menu.py", line 1, in <module>
    from archinstall.lib.global_menu import GlobalMenu
  File "/usr/lib/python3.12/site-packages/archinstall/__init__.py", line 282, in <module>
    post_process_arguments(arguments)
  File "/usr/lib/python3.12/site-packages/archinstall/__init__.py", line 275, in post_process_arguments
    load_plugin(path)
  File "/usr/lib/python3.12/site-packages/archinstall/lib/plugins.py", line 95, in load_plugin
    localized = _localize_path(path)
                ^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/site-packages/archinstall/lib/plugins.py", line 38, in _localize_path
    converted_path = Path(f'/tmp/{path.stem}_{hashlib.md5(os.urandom(12)).hexdigest()}.py')
                                  ^^^^^^^^^
AttributeError: 'str' object has no attribute 'stem'
```

Seems that the problem is in function `post_process_arguments` in file `archinstall/__init__.py`. This function passes a string to the function `load_plugin` instead of a 'Path' object.
I changed line 276 to `path = Path(args['plugin'])` and the problem was solved.

#Second Problem #
At this point new problem came out: 

```
Loading plugin from url ParseResult(scheme='https', netloc='', path='/raw.githubusercontent.com/phisch/archinstall-aur/refs/heads/master/archinstall-aur.py', params='', query='', fragment='')
Traceback (most recent call last):
  File "/home/archi/MyArchIso/install.py", line 3, in <module>
    from my_global_menu import Menu
  File "/home/archi/MyArchIso/my_global_menu.py", line 1, in <module>
    from archinstall.lib.global_menu import GlobalMenu
  File "/usr/lib/python3.12/site-packages/archinstall/__init__.py", line 282, in <module>
    post_process_arguments(arguments)
  File "/usr/lib/python3.12/site-packages/archinstall/__init__.py", line 275, in post_process_arguments
    load_plugin(path)
  File "/usr/lib/python3.12/site-packages/archinstall/lib/plugins.py", line 95, in load_plugin
    localized = _localize_path(path)
                ^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/site-packages/archinstall/lib/plugins.py", line 41, in _localize_path
    temp_file.write(urllib.request.urlopen(url.geturl()).read().decode('utf-8'))
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/urllib/request.py", line 215, in urlopen
    return opener.open(url, data, timeout)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/urllib/request.py", line 512, in open
    req = meth(req)
          ^^^^^^^^^
  File "/usr/lib/python3.12/urllib/request.py", line 1268, in do_request_
    raise URLError('no host given')
urllib.error.URLError: <urlopen error no host given>
```

The problem seems localized in funcion `_localize_path` in file `archinstall/lib/plugins.py`,  where funcion url.geturl() return "https:///raw.githubusercontent.com/phisch/archinstall-aur/master/archinstall-aur.py" (there are 3 backslash after "https:").

**Comment by shadowe1ite** [comment]
Any updates. I am facing the same issue. Please fix this

**Comment by systemofapwne** [comment]
The whole plugin systems seems to be deeply broken right now. I tried to write a simple plugin like
```python
class Plugin:
  def on_add_bootloader(self, *args, **kwargs):
    return True
```
to skip the `add_bootloader` stage for my custom install (Unfortunately `--skip-boot` is not doing the trick, so I opted for a plugin).
This causes import errors:

```
Traceback (most recent call last):
  File "/usr/bin/archinstall", line 5, in <module>
    from archinstall import run_as_a_module
  File "/usr/lib/python3.13/site-packages/archinstall/__init__.py", line 9, in <module>
    from archinstall.lib.args import arch_config_handler
  File "/usr/lib/python3.13/site-packages/archinstall/lib/args.py", line 549, in <module>
    arch_config_handler: ArchConfigHandler = ArchConfigHandler()
                                             ~~~~~~~~~~~~~~~~~^^
  File "/usr/lib/python3.13/site-packages/archinstall/lib/args.py", line 253, in __init__
    args: Arguments = self._parse_args()
                      ~~~~~~~~~~~~~~~~^^
  File "/usr/lib/python3.13/site-packages/archinstall/lib/args.py", line 439, in _parse_args
    load_plugin(plugin_path)
    ~~~~~~~~~~~^^^^^^^^^^^^^
  File "/usr/lib/python3.13/site-packages/archinstall/lib/plugins.py", line 91, in load_plugin
    from .args import arch_config_handler
ImportError: cannot import name 'arch_config_handler' from partially initialized module 'archinstall.lib.args' (most likely due to a circular import) (/usr/lib/python3.13/site-packages/archinstall/lib/args.py)
```

Adding imports to my plugin that other plugins seem to use, creates another `circular import` error

```python
import archinstall
from archinstall import Installer

class Plugin:
  def on_add_bootloader(self, *args, **kwargs):
    return True
```
throws
```
Loading plugin from url ParseResult(scheme='', netloc='', path='my-boot.py', params='', query='', fragment='')
Error: cannot import name 'Installer' from partially initialized module 'archinstall' (most likely due to a circular import) (/usr/lib/python3.13/site-packages/archinstall/__init__.py) The above error was detected when loading the plugin: my-boot.py
```

I then tried to import existing plugins via https, just to rule out me being stupid. I then found your issue. I fear, even when your `urlopen error no host given` error is getting fixed, you will stumble upon my issue.

**TL;DR:** The plugin system is broken

**Comment by anden04** [comment]
Any update on this? I'm experiencing the same issue.

**Comment by h8d13** [comment]
My guess is that a way to handle plugins would simply to add the hooks in `guided.py` or whichever type of script from the lib you use

```python
		if config.auth_config and config.auth_config.users:
			installation.create_users(
				config.auth_config.users,
				config.auth_config.privilege_escalation,
			)
			ShellApp().install(installation, config.auth_config.users)
```
This step also let's me clone repos per user into the target

The plugin system also pollutes readability of most of the `installer.py` file (which is already complex enough) AFAIK anything more major in terms of changes also has to be reflected there either way (or in related files) 

So having this 20 times in the code, I believe doesn't help many ppl 
```python
		for plugin in plugins.values():
```

**Comment by codefiles** [comment]
> The plugin system also pollutes readability of most of the `installer.py` file (which is already complex enough) AFAIK anything more major in terms of changes also has to be reflected there either way (or in related files)
> 
> So having this 20 times in the code, I believe doesn't help many ppl
> ```python
> 		for plugin in plugins.values():
> ```

I agree with this.

The plugin system has been broken for a significant duration of time and has not been maintained. Neither of the [reference implementations](https://archinstall.archlinux.page/archinstall/plugins.html#writing-your-own) have been active/updated in years.

@Torxed, would you be ok with the plugin system being removed?

**Comment by svartkanin** [comment]
Considering it is broken removing sounds fine as that will not have any immediate impact.

I guess another discussion would be what the purpose of it used to be and what use case it intended to solve. 

I think the approach of the plugin hooks sprinkled across the codebase made it somewhat unpredictable in anyways and maybe users should just be able to write post-installation scripts instead?

**Comment by codefiles** [comment]
Anything the plugin system was capable of should be scriptable anyway.

**Comment by hariomphulre** [comment]
hello @codefiles can you give me the previous archinstall/__init__.py code

**Comment by ScarletEmanu** [comment]
Hi. To me, the actual main issue is that the plugin system can't handle both URLs and simple paths. I don't know if there are any other bug in the plugin system. I can submit a PR to repair this.

**Comment by hariomphulre** [comment]
I submitted a PR on that issue

**Comment by ScarletEmanu** [comment]
I didn't see that, sorry

**Comment by hariomphulre** [comment]
#4524 plz review it

**Comment by ScarletEmanu** [comment]
I have fixed the plugin system. Let me know if there are still some bugs in it
